src/charge/modifiers/Once.php

```diff
@@ -0,0 +1,141 @@
+<?php declare(strict_types=1);
+
+namespace hiqdev\php\billing\charge\modifiers;
+
+use DateTimeImmutable;
+use hiqdev\php\billing\action\ActionInterface;
+use hiqdev\php\billing\charge\ChargeInterface;
+use hiqdev\php\billing\charge\modifiers\addons\MonthPeriod;
+use hiqdev\php\billing\charge\modifiers\addons\Period;
+use hiqdev\php\billing\charge\modifiers\addons\YearPeriod;
+use hiqdev\php\billing\charge\modifiers\exception\OnceException;
+
+/**
+ * 1. API:
+ * - once.per('1 year') – bill every month that matches the month of sale of the object
+ * - once.per('3 months') – bill every third month, starting from the month of sale of the object
+ * - once.per('day'), once.per('1.5 months') – throws an interpret-time exception, a value must NOT be a fraction of the month
+ * 2. In months where the formula should NOT bill, it should produce a ZERO charge.
+ * 3. If the sale is re-opened, the formula starts over.
+ * 4. It should throw exception for overuse prices.
+ */
+class Once extends Modifier
+{
+    private const MONTHS_IN_YEAR = 12;
+
+    public function per(string $interval): self
+    {
+        return $this->addAddon('per', $this->createInterval($interval));
+    }
+
+    private function createInterval(string $interval): Period
+    {
+        if ($this->isInvalidInterval($interval)) {
+            throw new OnceException("The interval cannot be a fraction.");
+        }
+
+        if ($this->isSupportedInterval($interval)) {
+            return Period::fromString($interval);
+        }
+
+        throw new OnceException("Invalid interval. Supported: whole months or years.");
```

Comment:
```suggestion
        $preiod = Period::fromString($interval);
        if (! $period instanceof MonthPeriod && ! $period instanceof YearPeriod) {
            throw new OnceException("Invalid interval. Supported: whole months or years.");
        }
        
        return $period;
```

Those Regexp checks in `isInvalidInterval` are redundant for me and break the incapsulation of `Preiod` class constraints.

Period class already checks for non-float values. Simply create a Period and check that it's OK to you.